### PR TITLE
Update ir_printer.cpp

### LINKED ISSUE #114035: Load model from jit script format. Repeating inference several times can lead to errors.
## 1. Save to jit script model from hf.
```
from transformers import BertForMaskedLM, BertTokenizer
import torch

model = BertForMaskedLM.from_pretrained('./data/Transformer-Bert', torchscript=True)
tokenizer = BertTokenizer.from_pretrained('./data/Transformer-Bert')
model.eval()
model.cuda()

with torch.no_grad():
    inp = ['[CLS] 中国的首都是哪里？ [SEP] 北京是 [MASK] 国的首都。 [SEP]']
    tokenizer_text = [tokenizer.tokenize(i) for i in inp]
    input_ids = [tokenizer.convert_tokens_to_ids(i) for i in tokenizer_text]
    input_ids = torch.LongTensor(input_ids)

    script_model = torch.jit.trace(model, input_ids.cuda())
    torch.jit.save(script_model, './data/Transformer-Bert/script_model.pt') 
```

## 2. Load jit script model and infer multi times.
```
from transformers import BertForMaskedLM, BertTokenizer
import torch

# Add following codes, will be ok.
"""
torch._C._jit_override_can_fuse_on_cpu(False)
torch._C._jit_override_can_fuse_on_gpu(False)
torch._C._jit_set_texpr_fuser_enabled(False)
torch._C._jit_set_nvfuser_enabled(False)
"""

tokenizer = BertTokenizer.from_pretrained('./data/Transformer-Bert')

with torch.no_grad():
    inp = ['[CLS] 中国的首都是哪里？ [SEP] 北京是 [MASK] 国的首都。 [SEP]']
    tokenizer_text = [tokenizer.tokenize(i) for i in inp]
    input_ids = [tokenizer.convert_tokens_to_ids(i) for i in tokenizer_text]
    input_ids = torch.LongTensor(input_ids)

model = torch.jit.load('./data/Transformer-Bert/script_model.pt')
model.eval()
out = model(input_ids.cuda())
print(out[0].shape)
out = model(input_ids.cuda())
print(out[0].shape)
out = model(input_ids.cuda())
print(out[0].shape)
```

## 3. Will have following exception:
```
torch.Size([1, 21, 21128])
torch.Size([1, 21, 21128])
Traceback (most recent call last):
  File "/root/open_grps_examples/py_examples/bert-torch/load_model.py", line 30, in <module>
    out = model(input_ids.cuda())
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: default_program(24): error: extra text after expected end of number

default_program(28): error: extra text after expected end of number

2 errors detected in the compilation of "default_program".

nvrtc compilation failed:

#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)


template<typename T>
__device__ T maximum(T a, T b) {
  return isnan(a) ? a : (a > b ? a : b);
}

template<typename T>
__device__ T minimum(T a, T b) {
  return isnan(a) ? a : (a < b ? a : b);
}

extern "C" __global__
void fused_mul_div_add(float* tattention_scores_63, float* tv_, float* aten_add, float* aten_mul) {
{
if (blockIdx.x<1ll ? 1 : 0) {
if ((long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x)<21ll ? 1 : 0) {
if (blockIdx.x<1ll ? 1 : 0) {
        float v = __ldg(tv_ + (long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x));
        aten_mul[(long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x)] = v * -3.402823466385289e+38.f;
      }    }  }if ((long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x)<5292ll ? 1 : 0) {
    float v_1 = __ldg(tattention_scores_63 + (long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x));
    float v_2 = __ldg(tv_ + ((long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x)) % 21ll);
    aten_add[(long long)(threadIdx.x) + 512ll * (long long)(blockIdx.x)] = v_1 / 8.f + v_2 * -3.402823466385289e+38.f;
  }}

```

### Versions

Collecting environment information...
PyTorch version: 2.1.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: Could not collect
CMake version: version 3.18.5
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-4.19.0-16-amd64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce GTX 1080 Ti
GPU 1: NVIDIA GeForce GTX 1080 Ti
GPU 2: NVIDIA GeForce GTX 1080 Ti
GPU 3: NVIDIA GeForce GTX 1080 Ti
GPU 4: NVIDIA GeForce GTX 1080 Ti
GPU 5: NVIDIA GeForce GTX 1080 Ti
GPU 6: NVIDIA GeForce GTX 1080 Ti
GPU 7: NVIDIA GeForce GTX 1080 Ti

Nvidia driver version: 525.105.17
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          56
On-line CPU(s) list:             0-55
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
CPU family:                      6
Model:                           79
Thread(s) per core:              2
Core(s) per socket:              14
Socket(s):                       2
Stepping:                        1
Frequency boost:                 enabled
CPU max MHz:                     2401.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        4799.92
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts flush_l1d
Virtualization:                  VT-x
L1d cache:                       896 KiB (28 instances)
L1i cache:                       896 KiB (28 instances)
L2 cache:                        7 MiB (28 instances)
L3 cache:                        70 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-13,28-41
NUMA node1 CPU(s):               14-27,42-55
Vulnerability Itlb multihit:     KVM: Mitigation: Split huge pages
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Meltdown:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] torch==2.1.1+cu118
[pip3] torchaudio==2.1.1+cu118
[pip3] torchvision==0.16.1+cu118
[pip3] triton==2.1.0
[conda] Could not collect

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### PR DESCRIPTION
**Pull request by pobin6**
Fixes #114035
when v larger than 1e16, the format will be error. example: v is 1.2e17, the output is 1.2e17.f, it have two point '.'

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/ir_printer.cpp [hunk 1]
```diff
@@ -191,25 +191,26 @@ void IRPrinter::visit(const CompareSelectPtr& v) {
...
-static void formatFPSuffix(std::ostream& os, T v) {
-  os << (v == std::ceil(v) ? ".f" : "f");
+static void formatFPSuffix(std::ostream& os, T v, bool flag) {
+  os << (flag && v == std::ceil(v) ? ".f" : "f");
 }
 
 template <typename T, std::enable_if_t<std::is_floating_point_v<T>>* = nullptr>
 static void formatImm(std::ostream& os, T v) {
   const int precision = 16;
+  const T const_range[] = {static_cast<T>(-1e16), static_cast<T>(1e16)};
```
Comment (EikanWang) [inline torch/csrc/jit/tensorexpr/ir_printer.cpp:206]: Do you think it is better to reuse precision?

```c++
  constexpr int precision = 16;
  constexpr T lower_bound = static_cast<T>(-std::pow(10, -precision));
  constexpr T upper_bound = -lower_bound ;
  const T const_range[] = {lower_bound, upper_bound};
```

File: torch/csrc/jit/tensorexpr/ir_printer.cpp [hunk 2]
```diff
@@ -191,25 +191,26 @@ void IRPrinter::visit(const CompareSelectPtr& v) {
   withParens(v->ret_val2());
 }
 
-static void formatFPSuffix(std::ostream& os, double v) {
-  os << (v == std::ceil(v) ? ".0" : "");
+static void formatFPSuffix(std::ostream& os, double v, bool flag) {
+  os << (flag && v == std::ceil(v) ? ".0" : "");
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:195]: @pobin6 , IIUC, the `std::setprecision(precision)` only impacts the output format but not change the value. It means that `v == std::ceil(v)` should return `False` if `v` is not a float number w/o fractional part.

File: torch/csrc/jit/tensorexpr/ir_printer.cpp (hunk 1 above)
Comment (pobin6) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:206]: It is better, but the second line,  `-std::pow(10, -precision)` should be `-std::pow(10, precision)`.

File: torch/csrc/jit/tensorexpr/ir_printer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pobin6) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:195]: Thank you for your explanation. Same as my thought.

**Review by jgong5 — CHANGES_REQUESTED** [review]
Could you please add a UT?

**Comment by pobin6** [comment]
@jgong5 Wrong click, don't mind. I can try, but it will take some time, unittest.

**Review by jgong5 — CHANGES_REQUESTED** [review]
Thanks. Please re-request review when you added UT.

**Comment by pobin6** [comment]
I add a unittest  at https://github.com/pytorch/pytorch/pull/140918
@jgong5

**Comment by jgong5** [comment]
> I add a unittest at #140918 @jgong5

Thanks. LGTM. But why not add the UT in this PR directly?

**Comment by pobin6** [comment]
Is it possible？ I also want to, but I haven't found the way to, not commonly used

**Review by jgong5 — COMMENTED** [review]
> Is it possible？ I also want to, but I haven't found the way to, not commonly used

I mean you can add the UT change from https://github.com/pytorch/pytorch/pull/140918 into this PR so that you have a bug fix together with a UT to cover it. It is a common practice. Any concern on this?

**Comment by pobin6** [comment]
I just find that's different branch😅

**Review by jgong5 — APPROVED** [review]

**Comment by prm-james-hill** [comment]
Following along because I also need this fix. Looks like ir_printer.cpp:205 needs to be marked as constexpr too.

File: torch/csrc/jit/tensorexpr/ir_printer.cpp [hunk 3]
```diff
@@ -191,25 +191,27 @@ void IRPrinter::visit(const CompareSelectPtr& v) {
...
-  os << (v == std::ceil(v) ? ".f" : "f");
+static void formatFPSuffix(std::ostream& os, T v, bool flag) {
+  os << (flag && v == std::ceil(v) ? ".f" : "f");
 }
 
 template <typename T, std::enable_if_t<std::is_floating_point_v<T>>* = nullptr>
 static void formatImm(std::ostream& os, T v) {
   const int precision = 16;
+  constexpr T lower_bound = static_cast<T>(-std::pow(10, precision));
+  constexpr T upper_bound = -lower_bound ;
```
Comment (prm-james-hill) [inline torch/csrc/jit/tensorexpr/ir_printer.cpp:207]: @pobin6 
```suggestion
  const T lower_bound = static_cast<T>(-std::pow(10, precision));
  const T upper_bound = -lower_bound ;
```

Comment (prm-james-hill) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:207]: The failures preventing this PR from being merged are all
`constexpr variable 'upper_bound' must be initialized by a constant expression`
(https://github.com/pytorch/pytorch/pull/140002#issuecomment-2462093346).

According to https://en.cppreference.com/w/cpp/numeric/math/pow, `std::pow` won't be `constexpr` until C++26. I don't know why your sample works. The failures are with Clang so maybe GCC has a non-standard extension

Comment (prm-james-hill) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:207]: Confirming the compiler difference:
```sh
james@work:~/code$ g++ main.cpp -std=c++11
james@work:~/code$ ./a.out
-10000000272564224.000000
james@work:~/code$ clang-19 main.cpp -std=c++11
main.cpp:3:17: error: constexpr variable 'lower_bound' must be initialized by a constant expression
    3 | constexpr float lower_bound = static_cast<float>(-std::pow(10, 16));
      |                 ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/cmath:1073:14: note: non-constexpr function 'pow' cannot be used in a constant expression
 1073 |       return pow(__type(__x), __type(__y));
      |              ^
main.cpp:3:51: note: in call to 'pow<int, int>(10, 16)'
    3 | constexpr float lower_bound = static_cast<float>(-std::pow(10, 16));
      |                                                   ^~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:140:17: note: declared here
  140 | __MATHCALL_VEC (pow,, (_Mdouble_ __x, _Mdouble_ __y));
      |                 ^
1 error generated.
```

**Comment by pobin6** [comment]
Thank your explanation, accepted. @prm-james-hill

**Comment by prm-james-hill** [comment]
@pobin6 I think you need to rebase this PR. The tests are failing again with:
```
##[error]Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/ec2-user/actions-runner/_work/pytorch/pytorch/.github/actions/upload-sccache-stats'. Did you forget to run actions/checkout before running your local action?
```
That GitHub Action exists at HEAD but not at this PR's base commit.

@jgong5 What is the process for getting this merged once the CI is passing?

**Comment by jgong5** [comment]
> @jgong5 What is the process for getting this merged once the CI is passing?

Ask pytorchbot to merge.

**Comment by prm-james-hill** [comment]
@pobin6 bumping and hoping you can finish merging this. I think all you need to do is:
1. Rebase to HEAD to fix CI. This PR is out of date so the GitHub Actions are failing
2. Write a comment saying "@pytorchbot merge" once CI is passing

**Comment by prm-james-hill** [comment]
> > @jgong5 What is the process for getting this merged once the CI is passing?
> 
> Ask pytorchbot to merge.

It looks like only pytorch collaborators can instruct the bot.

**Comment by pobin6** [comment]
a big miss, look like wrong command executed. @prm-james-hill 
```
git rebase main
git push --force
```
this PR look like no value to commit, so I decided to close it. if you want fixed, you can take a new PR.

**Comment by prm-james-hill** [comment]
> a big miss, look like wrong command executed. @prm-james-hill
> 
> ```
> git rebase main
> git push --force
> ```
> 
> this PR look like no value to commit, so I decided to close it. if you want fixed, you can take a new PR.

Recreated in #142119

**Pull request closed without merging**